atom/browser/api/atom_api_protocol.cc

```diff
@@ -202,42 +202,89 @@ Protocol::JsProtocolHandler Protocol::GetProtocolHandler(
+                                  base::Bind(&Protocol::RegisterProtocolInIO,
+                                             base::Unretained(this), scheme),
+                                  base::Bind(callback, v8::Null(isolate())));
+}
+
+void Protocol::OnInterceptProtocol(const std::string& scheme,
+                                   const JsProtocolHandler& handler,
+                                   const JsCompletionCallback& callback,
+                                   int is_handled) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  v8::Locker locker(isolate());
+  v8::HandleScope handle_scope(isolate());
+  if (!is_handled) {
+    callback.Run(v8::Exception::Error(
+        mate::StringToV8(isolate(), "Scheme does not exist.")));
+    return;
+  }
+
+  if (ContainsKey(protocol_handlers_, scheme)) {
+    callback.Run(v8::Exception::Error(
+        mate::StringToV8(isolate(), "Cannot intercept custom protocols.")));
+    return;
+  }
+
+  protocol_handlers_[scheme] = handler;
+  BrowserThread::PostTaskAndReply(BrowserThread::IO, FROM_HERE,
+                                  base::Bind(&Protocol::InterceptProtocolInIO,
+                                             base::Unretained(this), scheme),
+                                  base::Bind(callback, v8::Null(isolate())));
+}
+
 mate::ObjectTemplateBuilder Protocol::GetObjectTemplateBuilder(
     v8::Isolate* isolate) {
   return mate::ObjectTemplateBuilder(isolate)
-      .SetMethod("registerProtocol", &Protocol::RegisterProtocol)
-      .SetMethod("unregisterProtocol", &Protocol::UnregisterProtocol)
+      .SetMethod("_registerProtocol", &Protocol::RegisterProtocol)
+      .SetMethod("_unregisterProtocol", &Protocol::UnregisterProtocol)
       .SetMethod("registerStandardSchemes", &Protocol::RegisterStandardSchemes)
       .SetMethod("isHandledProtocol", &Protocol::IsHandledProtocol)
-      .SetMethod("interceptProtocol", &Protocol::InterceptProtocol)
-      .SetMethod("uninterceptProtocol", &Protocol::UninterceptProtocol);
+      .SetMethod("_interceptProtocol", &Protocol::InterceptProtocol)
+      .SetMethod("_uninterceptProtocol", &Protocol::UninterceptProtocol);
 }
 
 void Protocol::RegisterProtocol(v8::Isolate* isolate,
                                 const std::string& scheme,
-                                const JsProtocolHandler& callback) {
-  if (ContainsKey(protocol_handlers_, scheme) ||
-      job_factory_->IsHandledProtocol(scheme))
-    return node::ThrowError(isolate, "The scheme is already registered");
-
-  protocol_handlers_[scheme] = callback;
-  BrowserThread::PostTask(BrowserThread::IO,
-                          FROM_HERE,
-                          base::Bind(&Protocol::RegisterProtocolInIO,
-                                     base::Unretained(this), scheme));
+                                const JsProtocolHandler& handler,
+                                const JsCompletionCallback& callback) {
+  IsHandledProtocol(scheme,
```

Comment:
We should checking whether protocol can be registered when trying to register it in IO thread, calling IsHandledProtocol would make call jump from UI => IO => UI => IO, which is quite inefficient.